baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -80,6 +124,58 @@ def stop(self) -> None:
         # `should_stop` is an attribute of `ConsumerMixin`
         self.should_stop = True
 
+        if self._timer:
+            self._timer.stop()
+
+        messages = self._read_and_empty_batch()
+
+        if messages:
+            logger.debug("Requeueing %i messages", len(messages))
+            message: kombu.Message
+            for message in messages:
+                if not message.acknowledged:
+                    message.requeue()
```

Comment:
The handler does not get notified when the server gets terminated, so this unfortunately needs to happen in the pump. Alternatively, we can also make the message handler aware of the lifecycle, but I think that would be a bit too much.

Reply:
You're right, I think you're doing the right thing here but I what tripped me up is the `if not message.acknowledged:`.

These messages are all just sitting in the batch, they haven't been handed off to the handler yet so none of them should be `acknowledged`. If they _are_ in a state where they could be then we wouldn't want to do this since that would meant the batch was being processed and we wouldn't want to re-queue messages that are currently being processed.  You might want to remove the `if not message.acknowledged:`?  I'm not 100% sure about that though...

Reply:
Maybe we can print a warning if a message was already acknowledged?

```py
            for message in messages:
                if not message.acknowledged:
                    message.requeue()
                else:
                    logger.warning("Message in unprocessed batch had already been acknowledged") 
```

Reply:
That seems reasonable...

Reply:
If I saw this message in my logs I'd be pretty confused about what I'm supposed to do with it. Given the conversation above and the fact we're emitting at warning level I think I'm understanding that this is indicative of something the application needs to do differently? If so, can the message give some more context about where this is happening, what it means, and what to do about it?

(This thinking is coming from many attempts at making [this error message](https://github.com/reddit/baseplate.py/blob/develop/baseplate/lib/metrics.py#L243-L248) actionable. I _hope_ we're at a good place there now. If you want to copy that pattern I can help you get the shortlink set up to point at a relevant answer in [the FAQ](https://baseplate.readthedocs.io/en/stable/guide/faq.html))